mlir/lib/Dialect/Vector/Transforms/VectorTransforms.cpp

```diff
@@ -1607,7 +1607,23 @@ struct ChainedReduction final : OpRewritePattern<vector::ReductionOp> {
   }
 };
 
-/// For vectors with either leading or trailing unit dim, replaces:
+FailureOr<VectorType> dropNonScalableUnitDimType(VectorType VT) {
+  VectorType newVT = VT;
+  int removed = 0;
+  auto shape = VT.getShape();
+  for (unsigned i = 0; i < shape.size(); i++) {
+    if (shape[i] == 1 && !VT.getScalableDims()[i]) {
+      newVT = VectorType::Builder(newVT).dropDim(i - removed);
```

Comment:
Can you move the `VectorType::Builder()` outside the loop, so in the loop you just call `builder.dropDim`? Right now, this is constructing a new vector type every time it removes a dimension.